tutorials/editor_basics/editor_events.md

```diff
@@ -2,172 +2,168 @@
+As explained in [Working with Text](working_with_text.md#safely-modifying-text-in-a-document), safe modifications to the Document must be in the context of a write action. 
+So although a method on the `Document` interface does the `String` insertion, the write action ensures a stable context.
 ```java
-public class EditorHandlerIllustration extends AnAction {
-    @Override
-    public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
-        final Editor editor = anActionEvent.getRequiredData(CommonDataKeys.EDITOR);
-        EditorActionManager actionManager = EditorActionManager.getInstance();
-        EditorActionHandler actionHandler = actionManager.getActionHandler(IdeActions.ACTION_EDITOR_CLONE_CARET_BELOW);
-    }
-    @Override
-    public void update(@NotNull final AnActionEvent anActionEvent) {
-        //...
-    }
+class MyTypedHandler implements TypedActionHandler {
+  @Override
+  public void execute(@NotNull Editor editor, char c, @NotNull DataContext dataContext) {
+    final Document document = editor.getDocument();
+    Project project = editor.getProject();
+    Runnable runnable = () -> document.insertString(0, "editor_basics\n");
+    WriteCommandAction.runWriteCommandAction(project, runnable);
+  }
 }
 ```
 
-### 3.2.4. Making *EditorActionHandler* execute actions
-
-To execute an action we need to call the `public final void execute(@NotNull Editor editor, @Nullable final Caret contextCaret, final DataContext dataContext)`
-method of the corresponding *EditorActionHandler*:
+### Registering a Custom TypedActionHandler
+To enable a custom implementation of `TypedActionHandler` to receive editor keystroke events, it must be registered to replace the existing typing handler.
+The registration is done through the [TypedAction](upsource:///platform/platform-api/src/com/intellij/openapi/editor/actionSystem/TypedAction.java) class.
 
+As is shown in the snippet below, the `EditorActionManager` is used to get access to the `TypedAction` class.
+The method `TypedAction.setupHandler()` is used to register the custom `MyTypedHandler` class:
 ```java
 public class EditorHandlerIllustration extends AnAction {
-    @Override
-    public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
-        final Editor editor = anActionEvent.getRequiredData(CommonDataKeys.EDITOR);
-        EditorActionManager actionManager = EditorActionManager.getInstance();
-        EditorActionHandler actionHandler = actionManager.getActionHandler(IdeActions.ACTION_EDITOR_CLONE_CARET_BELOW);
-        actionHandler.execute(editor, editor.getCaretModel().getCurrentCaret(), anActionEvent.getDataContext());
-    }
-    @Override
-    public void update(@NotNull final AnActionEvent anActionEvent) {
-        //
+    static {
+        final EditorActionManager actionManager = EditorActionManager.getInstance();
```

Comment:
recently deprecated,  replaced to `com.intellij.openapi.editor.actionSystem.TypedAction#getInstance`

Reply:
Not fully in 2019.2. Will pick up when rolled into next release.